cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,250 @@
+    rolling_hash_ = (rolling_hash_ << 1) + GEARHASH_TABLE[nth_run_][bytes[i]];
+    has_matched_ = has_matched_ || ((rolling_hash_ & hash_mask_) == 0);
+  }
+}
+
+void ContentDefinedChunker::Roll(std::string_view value) {
+  chunk_size_ += value.size();
+  if (chunk_size_ < min_size_) {
+    // short-circuit if we haven't reached the minimum chunk size, this speeds up the
+    // chunking process since the gearhash doesn't need to be updated
+    return;
+  }
+  for (char c : value) {
+    rolling_hash_ =
+        (rolling_hash_ << 1) + GEARHASH_TABLE[nth_run_][static_cast<uint8_t>(c)];
+    has_matched_ = has_matched_ || ((rolling_hash_ & hash_mask_) == 0);
+  }
+}
+
+bool ContentDefinedChunker::NeedNewChunk() {
+  // decide whether to create a new chunk based on the rolling hash; has_matched_ is
+  // set to true if we encountered a match since the last NeedNewChunk() call
+  if (ARROW_PREDICT_FALSE(has_matched_)) {
+    has_matched_ = false;
+    // in order to have a normal distribution of chunk sizes, we only create a new chunk
+    // if the adjused mask matches the rolling hash 8 times in a row, each run uses a
+    // different gearhash table (gearhash's chunk size has exponential distribution, and
+    // we use central limit theorem to approximate normal distribution)
+    if (ARROW_PREDICT_FALSE(++nth_run_ >= 7)) {
+      nth_run_ = 0;
+      chunk_size_ = 0;
+      return true;
+    }
+  }
+  if (ARROW_PREDICT_FALSE(chunk_size_ >= max_size_)) {
+    // we have a hard limit on the maximum chunk size, not that we don't reset the rolling
+    // hash state here, so the next NeedNewChunk() call will continue from the current
+    // state
+    chunk_size_ = 0;
+    return true;
+  }
+  return false;
+}
+
+template <typename T>
+const std::vector<Chunk> ContentDefinedChunker::Calculate(const int16_t* def_levels,
+                                                          const int16_t* rep_levels,
+                                                          int64_t num_levels,
+                                                          const T& leaf_array) {
+  std::vector<Chunk> chunks;
+  bool has_def_levels = level_info_.def_level > 0;
+  bool has_rep_levels = level_info_.rep_level > 0;
+
+  if (!has_rep_levels && !has_def_levels) {
+    // fastest path for non-nested non-null data
+    int64_t offset = 0;
+    int64_t prev_offset = 0;
+    while (offset < num_levels) {
+      Roll(leaf_array.GetView(offset));
+      ++offset;
+      if (NeedNewChunk()) {
+        chunks.emplace_back(prev_offset, prev_offset, offset - prev_offset);
+        prev_offset = offset;
+      }
+    }
+    if (prev_offset < num_levels) {
+      chunks.emplace_back(prev_offset, prev_offset, num_levels - prev_offset);
+    }
+  } else if (!has_rep_levels) {
+    // non-nested data with nulls
+    int64_t offset = 0;
+    int64_t prev_offset = 0;
+    while (offset < num_levels) {
+      Roll(def_levels[offset]);
+      Roll(leaf_array.GetView(offset));
```

Comment:
Er, you shouldn't call `GetView` if the array element is null, because the raw value is undefined (and won't be written out to Parquet anyway).

Reply:
Fixed it, this possibly improves the performance as well.